haystack/utils/base_serialization.py

```diff
@@ -172,69 +177,98 @@ def _convert_to_basic_types(value: Any) -> Any:
+    Deserializes a value with schema information back to its original form.
 
     Takes a dict of the form:
       {
-         "schema": {
-            "numbers": {"type": "integer"},
-            "messages": {"type": "array", "items": {"type": "haystack.dataclasses.chat_message.ChatMessage"}},
-        },
-        "data": {
-            "numbers": 1,
-            "messages": [{"role": "user", "meta": {}, "name": None, "content": [{"text": "Hello, world!"}]}],
+         "serialization_schema": {"type": "integer"} or {"type": "object", "properties": {...}},
+         "serialized_data": <the actual data>
       }
 
     :param serialized: The serialized dict with schema and data.
-    :returns: The deserialized dict with original values.
+    :returns: The deserialized value in its original form.
     """
     schema = serialized.get("serialization_schema", {})
-    data = serialized.get("serialized_data", {})
-
-    result: Dict[str, Any] = {}
-    for field, raw in data.items():
-        info = schema.get(field)
-        # no schema entry → just deep-deserialize whatever we have
-        if not info:
-            result[field] = _deserialize_value(raw)
-            continue
-
-        t = info["type"]
+    data = serialized.get("serialized_data")
+
+    if not schema or "type" not in schema:
+        # No schema info, try to deserialize as-is
+        return _deserialize_value(data)
+
+    schema_type = schema["type"]
+
+    # Handle object case (dictionary with properties)
+    if schema_type == "object":
+        properties = schema.get("properties")
+        if properties:
+            result: Dict[str, Any] = {}
+
+            if isinstance(data, dict):
+                for field, raw_value in data.items():
+                    field_schema = properties.get(field)
+                    if field_schema:
+                        # Recursively deserialize each field - avoid creating temporary dict
+                        result[field] = _deserialize_value_with_schema(
+                            {"serialization_schema": field_schema, "serialized_data": raw_value}
+                        )
+                    else:
+                        # No schema for this field, deserialize as-is
+                        result[field] = _deserialize_value(raw_value)
+            return result
+        else:
+            # Old format: generic object
+            return _deserialize_value({"type": "object", "data": data})
+
+    # Handle array case
+    elif schema_type == "array":
+        if not isinstance(data, list):
+            return data
+
+        # Cache frequently accessed schema properties
+        item_schema = schema.get("items", {})
+        item_type = item_schema.get("type", "any")
+        is_set = schema.get("uniqueItems") is True
+        min_items = schema.get("minItems")
+        max_items = schema.get("maxItems")
+        is_tuple = min_items is not None and max_items is not None
```

Comment:
slight nit this could be one line since it doesn't look like we use minItems or maxItems anywhere else
```suggestion
        is_tuple = schema.get("minItems") is not None and schema.get("maxItems") is not None
```